### PR TITLE
Improving error message in DataIO

### LINKED ISSUE #6673: Improve Error Messages in DataIO.h
### Description of the enhancement or error report

Right now if you attempt to checkpoint or recover, you might receive an error message with no detail about raw pointers. It may not be obvious where to look for the issue.

### Rationale for the enhancement or information for reproducing the error

Create a custom type that has a raw pointer and run either `Outputs/checkpoint=true` or `--recover`

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Improved error message. No impact to any applications.

### PR DESCRIPTION
**Pull request by permcody**
This PR adds "demangle" to print a better error message when a raw pointer is seen in the checkpoint operation. It required me to split out the `Backup` type into it's own file due to both `DataIO` and `RestartableDataIO` needing access to that type.

closes #6673

### CODE REVIEW

**Comment by friedmud** [comment]
Weird failure... when you get that worked our this is meegeable

**Comment by permcody** [comment]
Yeah it's working on my system but wow! I didn't feel like digging into it
more so it'll sit for a few days.

File: framework/src/restart/Backup.C [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+// Backup Definitions
+Backup::Backup()
+{
+  unsigned int n_threads = libMesh::n_threads();
+
+  _restartable_data.resize(n_threads);
+
+  for (unsigned int i=0; i < n_threads; i++)
```
Comment (andrsd) [inline framework/src/restart/Backup.C]: Missing spaces around `=`

File: framework/src/restart/Backup.C [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    _restartable_data[i] = new std::stringstream;
+}
+
+Backup::~Backup()
+{
+  unsigned int n_threads = libMesh::n_threads();
+
+  for (unsigned int i=0; i < n_threads; i++)
```
Comment (andrsd) [inline framework/src/restart/Backup.C]: Same here

**Pull request merged**